app/src/main/java/ch/epfl/sdp/contamination/ConcreteAnalysis.java

```diff
@@ -101,28 +99,37 @@ private void modelInfectionEvolution(Map<Carrier, Integer> suspectedContacts) {
         return contactDuration;
     }
 
+    private float getFactor(int recoveryCounter){
+        return (float)(Math.pow(InfectionAnalyst.IMMUNITY_FACTOR,recoveryCounter) * TRANSMISSION_FACTOR) ;
+    }
+
+
     @Override
     public void updateInfectionPredictions(Location location, Date startTime, Callback<Void> callback) {
 
         Date now = new Date(System.currentTimeMillis());
-        receiver.getUserNearbyDuring(location, startTime, now, aroundMe -> {
-            modelInfectionEvolution(identifySuspectContacts(aroundMe));
-            callback.onCallback(null);
-        });
-
-        //Method 1 : (synchrone)
-//        int badMeetings = receiver.getAndResetSickNeighbors(me.getUniqueId());
-//        updateCarrierInfectionProbability(me.getIllnessProbability() + badMeetings*TRANSMISSION_FACTOR);
-
-        //method 2 : (asynchrone)
-        receiver.getNumberOfSickNeighbors(me.getUniqueId(), res -> {
-            float badMeetings = 0;
-            if(!((Map)(res)).isEmpty()){
-                badMeetings =  ((float) (((HashMap) (res)).get(publicAlertAttribute)));
+        receiver.getRecoveryCounter(me.getUniqueId(), recoveryCounter->{
+            int recoveryCounter1 = 0;
+            if(!((Map)(recoveryCounter)).isEmpty()){
+                recoveryCounter1 =  ((int) (((HashMap) (recoveryCounter)).get(privateRecoveryCounter)));
             }
-            updateCarrierInfectionProbability(Math.min(me.getIllnessProbability() + badMeetings * TRANSMISSION_FACTOR,1f));
-            cachedSender.resetSickAlerts(me.getUniqueId());
+            final int counter = recoveryCounter1;
+            receiver.getUserNearbyDuring(location, startTime, now, aroundMe -> {
+                modelInfectionEvolution(identifySuspectContacts(aroundMe),counter);
+                receiver.getNumberOfSickNeighbors(me.getUniqueId(), res -> {
+                    float badMeetings = 0;
+                    if (!((Map) (res)).isEmpty()) {
+                        badMeetings = ((float) (((HashMap) (res)).get(publicAlertAttribute)));
+                    }
+                    if(badMeetings!=0){
+                        updateCarrierInfectionProbability(Math.min(me.getIllnessProbability() + badMeetings * getFactor(counter), 1f));
+                        cachedSender.resetSickAlerts(me.getUniqueId());
+                    }
+                    callback.onCallback(null);
+                });
+            });
         });
```

Comment:
There are several lambdas here: maybe you could find some method to extract?

Reply:
Humm yeah but I can't extract it since I have to make one query one after the other. In fact I think that Kevin's PR will bring FUtures and will make all of this seem much cleaner.